src/output/plugins/visualization/SoundAnalysis.hxx

```diff
@@ -0,0 +1,317 @@
+	 * lo_cutoff. A typical value would be 10000-12000 (empirically, there's
+	 * not a lot of activity above 10000 in song data). Units
+	 * Hz. Configuration value "hi_cutoff" */
+	float hi_cutoff;
+
+	static constexpr size_t DEFAULT_NUM_SAMPLES = 513;
+	static constexpr size_t DEFAULT_LO_CUTOFF = 200;
+	static constexpr size_t DEFAULT_HI_CUTOFF = 10000;
+
+public:
+	SoundAnalysisParameters() noexcept;
+	explicit SoundAnalysisParameters(const ConfigBlock &config_block);
+	SoundAnalysisParameters(size_t num_samples, float lo_cutoff,
+				float hi_cutoff);
+
+	size_t
+	GetNumSamples() const noexcept {
+		return num_samples;
+	}
+	float
+	GetLoCutoff() const noexcept {
+		return lo_cutoff;
+	}
+	float
+	GetHiCutoff() const noexcept {
+		return hi_cutoff;
+	}
+};
+
+/**
+ * \class SoundAnalysis
+ *
+ * \brief Analayze PCM data in a manner convienient for visualization authors
+ *
+ *
+ * This class houses our logic for going from raw PCM data to the power
+ * spectrum, bass/mids/trebs &c. Instances are constructed with configuration
+ * information on the analysis details, and repeated analysis for different
+ * datasets is performed via update(). Since instances allocate input & output
+ * buffers for the discrete Fourier transform, they are not meant to be copied
+ * around.
+ *
+ *
+ */
+
+class SoundAnalysis {
+
+	/// # of samples to be used in each analysis
+	size_t num_samples;
+	/* # of Fourier coefficients computed by FFTW (should be
+	 * (num_samples / 2) + 1) */
+	size_t out_samples;
+	std::shared_ptr<SoundInfoCache> pcache;
+	AudioFormat audio_format;
+	/* # of audio channels (e.g. 1 is mono, 2 is stereo--
+	 * # cf. SampleFormat.hxx); should be audio_format.num_channels */
+	uint8_t num_channels;
+	/// Size of `buf`, in bytes
+	size_t cbbuf;
+	/// Pre-allocated buffer for raw PCM data
+	std::unique_ptr<std::byte[]> buf;
+	/// Input array for all FFTs performed by this `SoundAnalysis` instance
+	std::unique_ptr<float, void (*)(void*)> in;
+	/// Output array for all FFTs performed by this `SoundAnalysis` instance
+	std::unique_ptr<fftwf_complex, void (*)(void*)> out;
+	/* Pre-computed (by fftw) information on the fastest way to compute the
+	 * Discrete Fourier Transform on the underlying hardware */
+	fftwf_plan plan;
+	/* Frequency cutoffs, in Hz; we'll return frequencies in the range
+	   [freq_lo, freq_hi] */
+	float freq_lo, freq_hi;
+	/* Indicies into `out` corresponding to the desired frequency range;
+	 * that range is indexed by [index_lo, index_hi) */
+	size_t idx_lo, idx_hi;
+	/// Indicies into `out` corresponding "mids" & "trebs"
+	size_t idx_mids, idx_trebs;
+	/* Bass/mids/trebs, laid-out as [bass, mids, trebs, bass, mids, trebs]
+	 * (i.e. 3 * num_channels floats) */
+	std::unique_ptr<float[]> bass_mids_trebs;
+
+public:
+	SoundAnalysis(const SoundAnalysisParameters &params,
+		      std::shared_ptr<SoundInfoCache> pc);
+
+	uint8_t
+	NumChan() const noexcept {
```

Comment:
`NumChan` is a very obscure method name. What does it do? Does it numerate Jackie Chan?
Hey, this public method name is shorter than the private field it accesses!
If you really need this method, why not `GetChannelCount()`?